plugins/speculation-rules/load.php

```diff
@@ -20,15 +20,65 @@
 	exit;
 }
 
-// Define the constant.
-if ( defined( 'SPECULATION_RULES_VERSION' ) ) {
-	return;
-}
+(
+	/**
+	 * Register this copy of the plugin among other potential copies embedded in plugins or themes.
+	 *
+	 * @param string  $global_var_name Global variable name for storing the plugin pending loading.
+	 * @param string  $version         Version.
+	 * @param Closure $load            Callback that loads the plugin.
+	 */
+	static function ( string $global_var_name, string $version, Closure $load ) {
+		if ( ! isset( $GLOBALS[ $global_var_name ] ) ) {
+			$bootstrap = static function () use ( $global_var_name ) {
+				if (
+					isset( $GLOBALS[ $global_var_name ]['load'], $GLOBALS[ $global_var_name ]['version'] )
+					&&
+					$GLOBALS[ $global_var_name ]['load'] instanceof Closure
+					&&
+					is_string( $GLOBALS[ $global_var_name ]['version'] )
+				) {
+					call_user_func( $GLOBALS[ $global_var_name ]['load'], $GLOBALS[ $global_var_name ]['version'] );
+					unset( $GLOBALS[ $global_var_name ] );
+				}
+			};
+
+			// Wait until after the plugins have loaded and the theme has loaded. The after_setup_theme action is used
+			// because it is the first action that fires once the theme is loaded.
+			add_action( 'after_setup_theme', $bootstrap, PHP_INT_MIN );
+		}
+
+		// Register this copy of the plugin.
+		if (
+			// Register this copy if none has been registered yet.
+			! isset( $GLOBALS[ $global_var_name ]['version'] )
+			||
+			// Or register this copy if the version greater than what is currently registered.
+			version_compare( $version, $GLOBALS[ $global_var_name ]['version'], '>' )
+			||
+			// Otherwise, register this copy if it is actually the one installed in the directory for plugins.
+			rtrim( WP_PLUGIN_DIR, '/' ) === dirname( __DIR__ )
```

Comment:
We may need to think about systems that install a copy of this as an MU Plugin. Should we allow a plugin with a newer version of the API to override the MU Plugin copy?

Reply:
I think so because only plugins in the `plugins` directory can be updated.

Reply:
Only if you're talking about updating from the .org plugins directory. Lots of platforms require updating via deployments where you would update the mu-plugin in similar ways to any other plugin. Regardless, you're probably right that we should leave handling of MU Plugins to the people managing those systems.

The scenario that came to mind was when a platform intentionally installed the Speculation Rules plugin as part of an MU plugin setup, but allowed some other plugin to be installed that happened to bundle a newer version of the feature. As someone managing a platform, I would want the MU Plugin version to take priority, but I could write code to ensure that happens and load it as part of whatever loader file I've placed in the mu-plugins directory.

Reply:
The other reason for why I think a regular plugin should take precedence over an mu-plugin is the same reason as I outlined in the description:

> This is because there is a Settings link in the plugin row actions, and when something is misbehaving with the plugin it would be confusing for them to report an issue for the plugin which is active but isn't actually loaded.

If someone has the Speculative Loading plugin installed and they can see it in their plugins list as active, then if something goes awry with the plugin and they report a bug to the support forum, they should be able to share the version as listed in the plugins list page. (Granted, they could look for the generator meta tag, but users would be more likely to use the version listed in the plugin list table when reporting a bug.)